In [19]:
import pandas as pd
from thefuzz import fuzz
from thefuzz import process

In [20]:
oms_data = pd.read_csv('../data/processed/oms_data.csv')
paytm_data = pd.read_csv("../data/processed/paytm_data.csv")

In [21]:
oms_data.head()

MPI           Transaction #                   Receipt #  \
0  1000000100541195  OPD/JNR/23-24/CS/78006      OPD/JNR/23-24/RC/24180   
1  1000000100991335  OPD/JNR/23-24/CS/78004  OPD/JNR/23-24/CON/RC/53825   
2  1000000101881322  OPD/JNR/23-24/CS/78003      OPD/JNR/23-24/RC/24179   
3  1000000101936174  OPD/JNR/23-24/CS/78001  OPD/JNR/23-24/CON/RC/53824   
4  1000000100755069  OPD/JNR/23-24/CS/77999      OPD/JNR/23-24/RC/24176   

         Type            Patient Name  Amount  From Advance  \
0  OP  Orders                Vandana    600.0           0.0   
1  OP Consult            Sai Vedansh    850.0           0.0   
2  OP  Orders                 Jiyaan    300.0           0.0   
3  OP Consult  Ayeza Nishwa Fathima    1000.0           0.0   
4  OP  Orders           B/O Shruthi     540.0           0.0   

   Amount after Advance/Package  Cash    Card  ...  Bank/Vendor Amount Due  \
0                         600.0     0   600.0  ...          NaN          0   
1                         850.0     0   850.0  ...          NaN          0   
2                         300.0     0     0.0  ...          upi          0   
3                        1000.0     0  1000.0  ...          NaN          0   
4                         540.0     0     0.0  ...          upi          0   

   Comments           User Payment Aggregator Transaction ID  \
0       NaN  Manjunatha A                              43144   
1       NaN  Manjunatha A                              43141   
2       NaN  Manjunatha A                                NaN   
3       NaN   Salman Khan                              43136   
4       NaN   Salman Khan                                NaN   

  Standardized DateTime  UnixTimestamp  UnixTimestampAmount  \
0        20231231234050   1.704046e+09    1704046250 600.00   
1        20231231230650   1.704044e+09    1704044210 850.00   
2        20231231210250   1.704037e+09    1704036770 300.00   
3        20231231194250   1.704032e+09   1704031970 1000.00   
4        20231231192950   1.704031e+09    1704031190 540.00   

           DateTimeAmount FinalAmount  
0   20231231234050 600.00       600.0  
1   20231231230650 850.00       850.0  
2   20231231210250 300.00       300.0  
3  20231231194250 1000.00      1000.0  
4   20231231192950 540.00       540.0  

[5 rows x 21 columns]

In [22]:
paytm_data.head()

Transaction_ID                      Order_ID  \
0  20231201010880000937176714792993815  2023120102461608140510820434   
1  20231201010910000937272200426767146                    ORD1538522   
2  20231201010810000937237335752680236                    ORD1538384   
3  20231201010930000937195176500533947  2023120103594308140610820434   
4  20231201010820000937194197378068590                    ORD1538371   

  Payment_Mode Issuing_Bank    POS_ID   External_Serial_No  \
0   DEBIT_CARD         HDFC  10820434  3120110745724962821   
1  CREDIT_CARD         AXIS  10986499  3120112495789276165   
2  CREDIT_CARD         HDFC  10986499  3120111849129772033   
3  CREDIT_CARD        ICICI  10820434  3120111078180630529   
4  CREDIT_CARD        ICICI  10986499  3120111061469625345   

   Credit/Debit_Card_Last_4_Digits  Bank_Transaction_ID  Settled_Amount  \
0                             6207                81405           388.0   
1                             4429                41374           600.0   
2                             5667                41371           150.0   
3                             9002                81406           226.0   
4                             9002                41370           850.0   

  Auth_Code Bank/Gateway Card_Scheme User_Expected_Credit_Date  \
0    085174         HFPP      MASTER           DEBIT_CARD:null   
1    127986         HFPP        VISA          CREDIT_CARD:null   
2    029535         HFPP      MASTER          CREDIT_CARD:null   
3    026348         HFPP        VISA          CREDIT_CARD:null   
4    251481         HFPP        VISA          CREDIT_CARD:null   

   Standardized DateTime  UnixTimestamp UnixTimestampAmount  \
0         20231201024543   1.701379e+09   1701378943 388.00   
1         20231201090508   1.701402e+09   1701401708 600.00   
2         20231201064637   1.701393e+09   1701393397 150.00   
3         20231201035905   1.701383e+09   1701383345 226.00   
4         20231201035511   1.701383e+09   1701383111 850.00   

          DateTimeAmount  FinalAmount  
0  20231201024543 388.00        388.0  
1  20231201090508 600.00        600.0  
2  20231201064637 150.00        150.0  
3  20231201035905 226.00        226.0  
4  20231201035511 850.00        850.0

In [23]:
oms_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6680 entries, 0 to 6679
Data columns (total 21 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   MPI                                6680 non-null   int64  
 1   Transaction #                      6680 non-null   object 
 2   Receipt #                          6680 non-null   object 
 3   Type                               6680 non-null   object 
 4   Patient Name                       6680 non-null   object 
 5   Amount                             6680 non-null   float64
 6   From Advance                       6680 non-null   float64
 7   Amount after Advance/Package       6680 non-null   float64
 8   Cash                               6680 non-null   int64  
 9   Card                               6680 non-null   float64
 10  CHQ/Wallet                         6680 non-null   float64
 11  Bank/Vendor                        3237 non-null   objec

In [24]:
paytm_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4652 entries, 0 to 4651
Data columns (total 18 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Transaction_ID                   4652 non-null   object 
 1   Order_ID                         4652 non-null   object 
 2   Payment_Mode                     4652 non-null   object 
 3   Issuing_Bank                     4652 non-null   object 
 4   POS_ID                           4652 non-null   int64  
 5   External_Serial_No               4652 non-null   int64  
 6   Credit/Debit_Card_Last_4_Digits  4652 non-null   int64  
 7   Bank_Transaction_ID              4652 non-null   int64  
 8   Settled_Amount                   4652 non-null   float64
 9   Auth_Code                        4652 non-null   object 
 10  Bank/Gateway                     4652 non-null   object 
 11  Card_Scheme                      4652 non-null   object 
 12  User_Expected_Credit

# Duplicate Bank_Transactions in Paytm_data 
- Upon analysis I have found that there are duplicate Bank_Transaction_ID in paytm_data but mostly they are on different date so that wont be a problem the below code checks that 
  

In [25]:
duplicate_bank_transaction_ids = set(paytm_data[paytm_data.duplicated(subset=['Bank_Transaction_ID'])]['Bank_Transaction_ID'])

In [26]:
for bank_transaction_id in duplicate_bank_transaction_ids:
    matched = paytm_data[paytm_data['Bank_Transaction_ID'] == bank_transaction_id]
    matched.reset_index(inplace=True)
    first_match_date = str(matched['Standardized DateTime'][0])[:8]
    for j in range(1,len(matched)):
        if(str(matched['Standardized DateTime'][j])[:8] == first_match_date):
            print("They have same date for Transaction ID : ",bank_transaction_id)

### 4652 records in Paytm_EDC and 6680 records in OMS.

- so we will try to match each record from Paytm_EDC with OMS. 
- The reason for having more records in OMS is because there are many ZERO bills generated.
- Also some of the payments are done by cash so those bills won't be present in Paytm_EDC records.
  

In [27]:
paytm_data["Matched"] = False
paytm_data["Receipt #"] = ''
paytm_data["Discrepancy"] = ''
paytm_data["MatchingTechnique"] = ''
oms_data["Matched"] = False
oms_data["Transaction_ID"] = ''
oms_data["Discrepancy"] = ''
oms_data["MatchingTechnique"] = ''


# Direct Matching 
### (using Bank Transcation ID in Paytm EDC with Payment Aggregator Transaction ID in OMS
 

In [28]:
matches_from_direct_matching = 0

for i in range(len(paytm_data)):
    match = oms_data[str(paytm_data['Bank_Transaction_ID'][i]) == oms_data['Payment Aggregator Transaction ID']]
    if not match.empty:
        # match.reset_index(inplace=True)
        for j, row in match.iterrows():
            if(str(match['Standardized DateTime'][j])[:8] == str(paytm_data['Standardized DateTime'][i])[:8]):
                paytm_data.loc[i,'Matched'] = True
                paytm_data.loc[i,'Receipt #'] = match['Receipt #'][j]
                paytm_data.loc[i,'MatchingTechnique'] = 'Direct'
                oms_data.loc[j,'Matched'] = True
                oms_data.loc[j,'Transaction_ID'] = paytm_data.loc[i,'Transaction_ID']
                oms_data.loc[j,'MatchingTechnique'] = 'Direct'
                matches_from_direct_matching+=1
                # All the payments in Paytm EDC are made by CreditCard / Debit card
                # Therefore, we can use the "Settled_Amount" column to check if the amounts match with the "Card" column amount in oms_data

                if(match['Card'][j] != paytm_data['Settled_Amount'][i]):
                    paytm_data['Discrepancy'][i] = "The amounts didn't match for Transaction ID : ",paytm_data['Bank_Transaction_ID'][i] , " The amounts are 'Amount after Advance/Package' : ",match['Amount after Advance/Package'][j]," Amount Due : ",match['Amount Due'][j] , " Settled Amount : ", paytm_data['Settled_Amount'][i]
                break

print("Matches from Direct Matching Technique : ",matches_from_direct_matching)

Matches from Direct Matching Technique :  2971


# Analysis from the Direct Matching Techinque

- 2971 records have been matched
- The timestamp difference for the matched pair from the oms records and paytm records is ranging from (-/+) 65 seconds
- So lets take the threshold as 90 seconds just to be sure that we include even the outlier cases.

![difference in timestamps for the direct matching case](../reports/figures/difference_in_timestamps_oms_vs_paytm.png)

# Fuzzy Matching
### (using "UnixTimestampAmount" column in Paytm EDC with "UnixTimestampAmount" column in OMS)

In [29]:
# Records which are not matched yet after Direct Matching Technique is applied
paytm_data_unmatched_records = paytm_data[~paytm_data['Matched']]

In [30]:
# Logic for Fuzzy matching

# Technique 1 : Levenstein Distance

# I have read this post for understanding how to use thefuzz library and which score to use when
# https://stackoverflow.com/questions/31806695/when-to-use-which-fuzz-function-to-compare-2-strings

count = 0
for index , row in paytm_data_unmatched_records.iterrows():
    # perform fuzzy matching on UnixTimestampAmount columns
    possible_matches = process.extractOne(row['UnixTimestampAmount'],oms_data['UnixTimestampAmount'], scorer=fuzz.token_sort_ratio)
    print(f"""
          paytm_data (Date) : {row["Standardized DateTime"]}
          paytm_data (Amount) : {row["UnixTimestampAmount"]}
          oms_data (Date) : {oms_data.loc[possible_matches[2],"Standardized DateTime"]}
          oms_data (Date Amount) : {oms_data.loc[possible_matches[2],"UnixTimestampAmount"]}
          """)
    if(count >= 10):
        break
    count+=1
    


          paytm_data (Date) : 20231201024543
          paytm_data (Amount) : 1701378943 388.00
          oms_data (Date) : 20231221111350
          oms_data (Date Amount) : 1703137430 780.00
          

          paytm_data (Date) : 20231201035905
          paytm_data (Amount) : 1701383345 226.00
          oms_data (Date) : 20231229094550
          oms_data (Date Amount) : 1703823350 650.00
          

          paytm_data (Date) : 20231201151420
          paytm_data (Amount) : 1701423860 23028.00
          oms_data (Date) : 20231212173450
          oms_data (Date Amount) : 1702382690 2320.00
          

          paytm_data (Date) : 20231201151349
          paytm_data (Amount) : 1701423829 2172.00
          oms_data (Date) : 20231212173450
          oms_data (Date Amount) : 1702382690 2320.00
          

          paytm_data (Date) : 20231201142158
          paytm_data (Amount) : 1701420718 412.00
          oms_data (Date) : 20231201142150
          oms_data (Date Amount) : 170142071

In [31]:
# Technique 1 : Levenstein Distance

# I have read this post for understanding how to use thefuzz library and which score to use when
# https://stackoverflow.com/questions/31806695/when-to-use-which-fuzz-function-to-compare-2-strings

count = 0

for index , row in paytm_data_unmatched_records.iterrows():
    # perform fuzzy matching on DateTimeAmount columns
    possible_matches = process.extractOne(row['DateTimeAmount'],oms_data['DateTimeAmount'], scorer=fuzz.token_sort_ratio)
    print(f"""
          paytm_data (Date) : {row["Standardized DateTime"]}
          paytm_data (Amount) : {row["DateTimeAmount"]}
          oms_data (Date) : {oms_data.loc[possible_matches[2],"Standardized DateTime"]}
          oms_data (Date Amount) : {oms_data.loc[possible_matches[2],"DateTimeAmount"]}
          """)

    if(count >= 10):
        break
    count+=1
    


          paytm_data (Date) : 20231201024543
          paytm_data (Amount) : 20231201024543 388.00
          oms_data (Date) : 20231207102450
          oms_data (Date Amount) : 20231207102450 310.00
          

          paytm_data (Date) : 20231201035905
          paytm_data (Amount) : 20231201035905 226.00
          oms_data (Date) : 20231217105050
          oms_data (Date Amount) : 20231217105050 226.00
          

          paytm_data (Date) : 20231201151420
          paytm_data (Amount) : 20231201151420 23028.00
          oms_data (Date) : 20231230151250
          oms_data (Date Amount) : 20231230151250 310.00
          

          paytm_data (Date) : 20231201151349
          paytm_data (Amount) : 20231201151349 2172.00
          oms_data (Date) : 20231201114950
          oms_data (Date Amount) : 20231201114950 700.00
          

          paytm_data (Date) : 20231201142158
          paytm_data (Amount) : 20231201142158 412.00
          oms_data (Date) : 20231201142150
          

# Analysis on using Fuzzy Matching using (Levenshtein Distance)

- The results are not good. if we just use levenshtein distance, we are not getting the correct matches.


# Tweaking the approach : Levenshtein Distance (on smaller range)
- Modifying the search space, by limiting the search only on the records which are close by in time. 

### UnixTimestampAmount

In [32]:
threshold = 90 # in seconds
fuzzySearchColumn = "UnixTimestampAmount"
count = 0
unique_ids = set()

for index , row in paytm_data_unmatched_records.iterrows():
    # Performing fuzzy matching on UnixTimestampAmount column and limit the search space only on the records which are close by in time based on the Threshold value and matching amounts only.
    possible_matches = process.extract(row[fuzzySearchColumn],oms_data[(abs(row['UnixTimestamp']-oms_data['UnixTimestamp'])<=threshold ) & (row["FinalAmount"]==oms_data['FinalAmount']) ][fuzzySearchColumn], scorer=fuzz.token_sort_ratio)
    if len(possible_matches):
        print(row[fuzzySearchColumn])
        print(possible_matches)
        count+=1
        
        for possible_match in possible_matches:
            unique_ids.add(possible_match[2])

print("Total Fuzzy Matches Found : ",count)
print("Unique IDs : ",len(unique_ids))


1701508807 1200.00
[('1701508730 1200.00', 89, 6219)]
1701501464 780.00
[('1701501410 780.00', 88, 6297), ('1701501530 780.00', 82, 6293)]
1701492341 1000.00
[('1701492350 1000.00', 89, 6420)]
1701674763 850.00
[('1701674690 850.00', 88, 5865), ('1701674690 850.00', 88, 5866)]
1701673189 850.00
[('1701673130 850.00', 88, 5886)]
1701690051 850.00
[('1701690110 850.00', 88, 5786)]
1701752792 1050.00
[('1701752870 1050.00', 89, 5680)]
1701846558 850.00
[('1701846590 850.00', 88, 5423)]
1701839056 850.00
[('1701839030 850.00', 88, 5497)]
1701917657 850.00
[('1701917630 850.00', 88, 5323)]
1701962122 850.00
[('1701962090 850.00', 82, 5138)]
1702110741 850.00
[('1702110770 850.00', 88, 4708)]
1702533780 3200.00
[('1702533770 3200.00', 94, 3718)]
1702571051 2320.00
[('1702571030 2320.00', 89, 3602)]
1702654472 850.00
[('1702654490 850.00', 88, 3447)]
1702708328 850.00
[('1702708250 850.00', 88, 3323)]
1702722872 1700.00
[('1702722950 1700.00', 83, 3205)]
1702888124 600.00
[('1702888130 600.00

### DateTimeAmount

In [33]:
threshold = 90 # in seconds
fuzzySearchColumn = "DateTimeAmount"
count = 0

for index , row in paytm_data_unmatched_records.iterrows():
    # Performing fuzzy matching on DateTimeAmount column and limit the search space only on the records which are close by in time based on the Threshold value and matching amounts only.
    possible_matches = process.extractBests(row[fuzzySearchColumn],oms_data[(abs(row['UnixTimestamp']-oms_data['UnixTimestamp'])<=threshold ) ][fuzzySearchColumn], scorer=fuzz.token_sort_ratio)
    if len(possible_matches) and (row['Settled_Amount'] == oms_data.loc[possible_matches[0][2],'Card'] or row['Settled_Amount'] == oms_data.loc[possible_matches[0][2],'CHQ/Wallet']):
        print(row[fuzzySearchColumn])
        print(possible_matches)
        count+=1

print("Total Fuzzy Matches Found : ",count)

20231202145007 1200.00
[('20231202144850 1200.00', 91, 6219), ('20231202145050 780.00', 79, 6218)]
20231202124744 780.00
[('20231202124850 780.00', 86, 6293), ('20231202124750 850.00', 86, 6294), ('20231202124750 700.00', 86, 6295), ('20231202124650 780.00', 86, 6297), ('20231202124650 850.00', 81, 6296)]
20231202101541 1000.00
[('20231202101550 1000.00', 91, 6420), ('20231202101650 780.00', 70, 6419)]
20231204125603 850.00
[('20231204125450 850.00', 90, 5865), ('20231204125450 850.00', 90, 5866)]
20231204122949 850.00
[('20231204122850 850.00', 86, 5886), ('20231204123050 1320.00', 70, 5885)]
20231204171051 850.00
[('20231204171150 850.00', 90, 5786)]
20231205103632 1050.00
[('20231205103750 1050.00', 86, 5680)]
20231206123918 850.00
[('20231206123950 850.00', 90, 5423), ('20231206123950 0.00', 85, 5424), ('20231206123850 1000.00', 79, 5425), ('20231206123750 900.00', 76, 5426)]
20231206103416 850.00
[('20231206103350 850.00', 86, 5497)]
20231207082417 850.00
[('20231207082350 850.00'

# Analysis on using Fuzzy Matching using (Levenshtein Distance) on smaller search space
- The results are much better when we limit the search space.
- We found a total of 35 matching records using Fuzzy Matching on smaller search space.

In [34]:
# Applying the Fuzzy Matching Technique and updating the entries in the dataframe

threshold = 90 # in seconds
fuzzySearchColumn = "UnixTimestampAmount"
matches_from_fuzzy_matching = 0

for index , row in paytm_data_unmatched_records.iterrows():
    # Performing fuzzy matching on UnixTimestampAmount column and limit the search space only on the records which are close by in time based on the Threshold value and matching amounts only.
    possible_matches = process.extractOne(row[fuzzySearchColumn],oms_data[(abs(row['UnixTimestamp']-oms_data['UnixTimestamp'])<=threshold ) & (row["FinalAmount"]==oms_data['FinalAmount']) & (oms_data['Matched']==False)][fuzzySearchColumn], scorer=fuzz.token_sort_ratio)
    # print((oms_data[(abs(row['UnixTimestamp']-oms_data['UnixTimestamp'])<=threshold ) & (row["FinalAmount"]==oms_data['FinalAmount']) & (oms_data['Matched']==False)][fuzzySearchColumn]))
    # print(possible_matches)
    if possible_matches:
        oms_data_index = possible_matches[2]
        # print(oms_data_index)
        paytm_data.loc[index,'Matched'] = True
        paytm_data.loc[index,'Receipt #'] = oms_data.loc[oms_data_index,'Receipt #']
        paytm_data.loc[index,'MatchingTechnique'] = 'Fuzzy'

        oms_data.loc[oms_data_index,'Matched'] = True
        oms_data.loc[oms_data_index,'Transaction_ID'] = paytm_data.loc[i,'Transaction_ID']
        oms_data.loc[oms_data_index,'MatchingTechnique'] = 'Fuzzy'
        matches_from_fuzzy_matching+=1

print("Total Fuzzy Matches Found : ",matches_from_fuzzy_matching)


Total Fuzzy Matches Found :  29


In [35]:
print(f"""
      {matches_from_direct_matching} Matches from Direct Matching Technique
      {matches_from_fuzzy_matching} Matches from Fuzzy Matching Technique
      """)



      2971 Matches from Direct Matching Technique
      29 Matches from Fuzzy Matching Technique
      


# Saving the results

In [36]:
paytm_data.to_csv("../reports/final_processed_paytm_data.csv", index=False)
oms_data.to_csv("../reports/final_processed_oms_data.csv", index=False)